File: server/bot/lib/integration/test/front_info_test.cpp
#include <fb/bot/integration/front_info_test.h>
#include <fb/bot/integration/game_controller.h>

using namespace std::chrono_literals;
using namespace fb::bot::integration;

front_info_test::front_info_test(game_bot_controller& controller) :
    bot_integration_test(controller, 2) // Use 2 bots
{ }

async::task<void> front_info_test::on_initialize(game_bot_controller& controller)
{
    auto bots = this->get_test_bots();

    auto& bot2 = bots[1];
    co_await bot2->move(DIRECTION::BOTTOM);
}

async::task<void> front_info_test::on_scenario_started(uint32_t scenario_index)
{
    co_return;
}

async::task<void> front_info_test::on_scenario_finished(uint32_t scenario_index)
{
    auto  bots = this->get_test_bots();
    auto& bot1 = bots[0];
    co_await bot1->clear_all_drop_items(DEFAULT_TIMEOUT);
    co_await this->sleep(DEFAULT_INTERVAL);
}

generator<bot_integration_test::scenario_t> front_info_test::on_generate_scenario()
{
    co_yield [this]() -> async::task<bool> {
        co_return co_await this->test_front_info();
    };
}

std::string front_info_test::name() const
{
    return "Front Info Test";
}

async::task<bool> front_info_test::test_front_info()
{
    fb::logger::debug("Starting front_info packet handler test");

    auto  bots = this->get_test_bots();
    auto& bot1 = bots[0];
    auto& bot2 = bots[1];

    bot2->chat("Creating items");
    co_await bot2->create_item("목도", 1, DEFAULT_TIMEOUT);
    co_await bot2->create_item("목검", 1, DEFAULT_TIMEOUT);
    co_await bot2->create_item("도토리", 100, DEFAULT_TIMEOUT);

    bot2->chat("Changing money to 10000");
    co_await bot2->change_money(10000, DEFAULT_TIMEOUT);

    bot2->chat("Dropping items and money");
    co_await bot2->drop_item(0, true, DEFAULT_TIMEOUT);
    co_await bot2->drop_item(1, true, DEFAULT_TIMEOUT);
    co_await bot2->drop_item(2, true, DEFAULT_TIMEOUT);
    co_await bot2->drop_money(10000, DEFAULT_TIMEOUT);

    bot1->chat("Sending front_info request");

    auto expected_messages =
        std::unordered_set<std::string>{bot2->name(), "목도", "목검", "도토리 100개", "금덩어리 10000전"};
    std::ignore = co_await bot1->request<fb::protocol::game::response::message>(
        fb::protocol::game::request::front_info(),
        [&expected_messages](auto& resp) -> bool {
            if (resp.type != MESSAGE_TYPE::STATE)
                return false;

            if (expected_messages.contains(resp.text))
                expected_messages.erase(resp.text);
            return expected_messages.empty();
        },
        DEFAULT_TIMEOUT);

    bot1->chat("Front info test completed");
    co_return true;
}

File: scripts/integration/front_info_test.lua
local lib      = require("integration.lib")
local resp     = require("integration.response")
local protocol = require("integration.protocol")

test_suite {
    name      = "Front Info Test",
    bot_count = 2,

    on_initialize = function(ctx)
        log("debug", "Front info test initialized")
        lib.formation.arrange_in_line(ctx)
    end,

    on_scenario_finished = function(ctx)
        ctx:bot(0):clear_all_drop_items()
    end,

    scenarios = {
        function(ctx)
            log("debug", "Starting front_info packet handler test")
            local bot1 = ctx:bot(0)
            local bot2 = ctx:bot(1)

            bot2:create_item("목도", 1)
            bot2:create_item("목검", 1)
            bot2:create_item("도토리", 100)
            bot2:money(10000)

            bot2:drop_item(0, true)
            bot2:drop_item(1, true)
            bot2:drop_item(2, true)
            bot2:drop_money(10000)

            local expected = {
                [bot2:name()] = true,
                ["목도"] = true,
                ["목검"] = true,
                ["도토리 100개"] = true,
                ["금덩어리 10000전"] = true,
            }
            bot1:direction("RIGHT")

            bot1:request(
                resp.message,
                protocol.front_info(),
                function(packet)
                    if packet.type ~= "STATE" then
                        return false
                    end
                    if expected[packet.text] then
                        expected[packet.text] = nil
                    end
                    local remaining = 0
                    for _ in pairs(expected) do
                        remaining = remaining + 1
                    end
                    return remaining == 0
                end)

            return true
        end,
    },
}
